File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.cpp
class Solution {
public:

    long long helper(string s, int i, long long result){
        if(i >= s.length() || !isdigit(s[i])){
            return result;
        }

        int digit = s[i] - '0';
        result = result * 10 + digit;
        if(result > INT_MAX) return result;
        return helper(s, i+1, result);

    }

    int myAtoi(string s) {
        int i = 0;
        bool flag = false;
        if(s.empty()) return 0;
        while(i < s.length() && s[i] == ' ') i++;

        s = s.substr(i);

        i =  0;
        if(s[i] == '-'){
            flag = true;
            i++;
        }else if(s[i] == '+'){
            i++;
        }

        long long res = helper(s, i, 0);
        if(flag == true){
            if(-res < INT_MIN) return INT_MIN;
            else return -res;
        }else{
            if(res > INT_MAX) return INT_MAX;
            else return res;
        }
    }
};

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        INT_MIN = -2**31
        INT_MAX = 2**31 - 1
        n = len(s)
        i = 0
        while i < n and s[i] == ' ':
            i+=1
        if i == n:
            return 0
        sign = 1
        if s[i] == '-':
            sign = -1
        if s[i] == '-' or s[i] == '+':
            i = i + 1
        num = 0
        while i < len(s):
            if s[i].isdigit() == False:
                break
            digit = int(s[i])
            num = num * 10 + digit
            if sign == 1 and num > INT_MAX:
                return INT_MAX
            if sign == -1 and num * -1 < INT_MIN:
                return INT_MIN
            i+=1
        return int(num * sign)

            